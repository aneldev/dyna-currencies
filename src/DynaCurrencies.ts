import {
  IDynaLabel,
  IDynaPrice
} from "dyna-interfaces";

import { currenciesInfo } from "./currenciesInfo";
import { countriesInfo } from "./countriesInfo";

export interface ICurrencyRates {
  [currencyName: string]: number | undefined;
}

export interface ICurrencies {
  [currencyName: string]: ICurrency;
}

export interface ICurrency {
  code: string;
  symbol: string;
  name: string;
  namePlural: string;
  symbolNative: string;
  decimalDigits: number;
  rounding: number;
}

export interface ICountries {
  [currencyName: string]: ICountry | undefined;
}

export interface ICountry {
  name: string;
  native: string;
  phone: string;
  continent: string;
  capital: string;
  currency: string;
  languages: string[];
}

export interface IDynaLabelCurrency extends IDynaLabel {
  values: {
    value: number;
    decimals: number;
    currencyName: string;
    currencyNamePlural: string;
    currencyCode: string;
    currencySymbol: string;
    currencySymbolNative: string;
  }
}

export class DynaCurrencies {
  private _currencyRates: ICurrencyRates = {};
  private _lastUpdate: Date | null = null;

  public updateRates(rates: ICurrencyRates): void {
    Object.keys(rates)
      .forEach((currencyName: string) => this._currencyRates[currencyName.toLowerCase()] = rates[currencyName]);
    this._lastUpdate = new Date;
  }

  public clearRates(): void {
    this._currencyRates = {};
    this._lastUpdate = null;
  }

  public get count(): number {
    return Object.keys(this._currencyRates).length;
  }

  public get lastUpdate(): Date | null {
    return this._lastUpdate;
  }

  public get hasRates(): boolean {
    return !!this.count;
  }

  public convert(value: number, fromCurrency: string, toCurrency: string, round: boolean = false): number | null {
    const fromCurrencyLC = fromCurrency.toLowerCase();
    const toCurrencyLC = toCurrency.toLowerCase();
    const toCurrencyUC = toCurrency.toUpperCase();

    if (fromCurrencyLC === toCurrencyLC) return value;

    const fromRate = this._currencyRates[fromCurrencyLC];
    const toRate = this._currencyRates[toCurrencyLC];
    const curInfo = currenciesInfo[toCurrencyUC];

    if (fromRate === undefined) return null;
    if (toRate === undefined) return null;
    if (curInfo === undefined) return null;

    let output: number = value * fromRate / toRate;

    if (round) {
      const factor: number = Math.pow(10, curInfo.decimalDigits);
      output = Math.round(output * factor) / factor;
    }

    return output;
  }

  public convertDynaPrice(price: IDynaPrice, toCurrency: string): IDynaPrice | null {
    const toCurrencyLC = toCurrency.toLowerCase();
    const value = this.convert(price.value, price.currency, toCurrencyLC);
    if (value === null) return null;
    return {
      value,
      currency: toCurrencyLC,
    };
  }

  public convertToLabel(value: number, fromCurrency: string, toCurrency: string): IDynaLabelCurrency | null {
    const currInfo = currenciesInfo[toCurrency.toUpperCase()] || {};
    const cValue = this.convert(value, fromCurrency, toCurrency);
    if (currInfo === undefined) return null;
    if (cValue === null) return null;

    const cValueString: number = Math.round(cValue * Math.pow(10, currInfo.decimalDigits || 0)) / Math.pow(10, currInfo.decimalDigits || 0);

    return {
      text: `${cValueString}${currInfo.symbol}`,
      values: {
        value: cValue,
        decimals: currInfo.decimalDigits,
        currencyName: currInfo.name,
        currencyNamePlural: currInfo.namePlural,
        currencyCode: currInfo.code,
        currencySymbol: currInfo.symbol,
        currencySymbolNative: currInfo.symbolNative,
      }
    }
  }

  public getCurrencyRatesDic(): ICurrencyRates {
    return {...this._currencyRates};
  }

  public getCurrencyRatesArray(): ICurrency[] {
    return Object.keys(currenciesInfo)
      .reduce((acc: ICurrency[], code: string) => {
        acc.push(currenciesInfo[code]);
        return acc;
      }, []);
  }

  public getCurrencyRatesByCountry(countryCode: string): ICurrency[] {
    const country = countriesInfo[countryCode.toUpperCase()];

    if (!country) return [];

    return country
      .currency
      .split(',')
      .map((currencyCode: string) => currenciesInfo[currencyCode])
      .filter(Boolean);
  }

  public getCurrencyByCountry(countryCode: string): ICurrency | null {
    const countryCurrencyRate = this.getCurrencyRatesByCountry(countryCode)[0];
    if (!countryCurrencyRate) return null;
    return countryCurrencyRate;
  }
}

