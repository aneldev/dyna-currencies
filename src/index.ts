import {
  DynaCurrencies,
  ICurrencyRates,
  ICurrency,
  IDynaLabelCurrency,
} from './DynaCurrencies';

export {
  DynaCurrencies,
  ICurrencyRates,
  ICurrency,
  IDynaLabelCurrency,
};
