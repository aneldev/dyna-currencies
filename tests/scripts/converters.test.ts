import "jest";

import {DynaCurrencies, IDynaLabelCurrency, ICurrencyRates, ICurrency} from '../../src/index';

const ratesText: string = `{"cad":46.1916,"tnd":23.3343,"rsd":0.570674,"bgn":34.6998,"isk":0.547281,"pab":57.8802,"pgk":18.0727,"iqd":0.0496036,"yer":0.23125,"bam":34.7771,"bnd":42.512,"brl":18.2717,"vnd":0.00254673,"sos":0.0996402,"huf":0.21786000000000003,"myr":13.6563,"szl":4.23543,"bbd":28.9401,"lrd":0.491978,"kpw":0.443319,"kgs":0.841167,"sll":0.00755799,"kmf":0.138257,"cop":0.019611,"mxn":3.17877,"uyu":1.98799,"aud":45.106,"std":0.00277178,"top":26.6704,"kyd":70.5856,"tzs":0.0257759,"lvl":96.7816,"srd":7.79029,"nad":4.23543,"dkk":9.12142,"mur":1.69321,"btn":0.88399,"bmd":57.8802,"php":1.1326,"ron":14.7845,"gyd":0.278449,"mmk":0.0424789,"gnf":0.00645576,"clp":0.0913881,"dzd":0.509336,"sgd":42.4225,"fjd":28.4566,"mnt":0.0234724,"kes":0.559771,"ngn":0.160793,"irr":0.00171303,"tjs":6.57344,"thb":1.73277,"ugx":0.0160575,"wst":23.1391,"shp":76.6419,"bdt":0.707676,"htg":0.917332,"try":16.1305,"mad":6.13102,"uah":2.16814,"twd":1.90176,"gip":76.6281,"gel":23.4152,"lsl":4.23543,"bsd":57.8802,"chf":59.4605,"jmd":0.448254,"all":0.507967,"ang":32.3512,"hkd":7.4003,"ern":3.75759,"csk":57.8702,"cup":2.18416,"vuv":0.537586,"gtq":7.88623,"aoa":0.348847,"lyd":42.2557,"pen":17.7167,"bwp":5.57528,"pln":15.7273,"lak":0.00696068,"djf":0.323918,"gmd":1.2256,"jod":81.6577,"afn":0.848065,"ils":16.403,"uzs":0.0071667,"eur":67.9076,"pyg":0.0102382,"kwd":191.541,"azn":34.0039,"mga":0.0190843,"bhd":153.436,"inr":0.8852150000000001,"sek":7.0712,"nok":7.22595,"awg":32.3353,"ttd":8.58375,"xpf":0.570061,"mro":0.162231,"gbp":76.9843,"sdg":8.66674,"usd":57.8134,"ghs":13.1223,"zmk":0.00595744,"omr":150.421,"xaf":0.103706,"byn":29.4321,"tmt":16.5181,"xdr":81.7077,"cve":0.61653,"zar":4.23641,"xof":0.103706,"rwf":0.0684434,"scr":4.22194,"pkr":0.549439,"mop":7.1948,"vef":5.68939,"npr":0.552454,"fkp":76.6419,"cdf":0.0369761,"idr":0.00428036,"lkr":0.377695,"mkd":1.10675,"jpy":0.511872,"egp":3.28025,"ars":3.32402,"mwk":0.0796816,"sbd":7.42699,"krw":0.0503829,"byr":0.00298291,"nio":1.90499,"syp":0.11227,"khr":0.0142788,"zwl":57.8802,"mzn":0.949043,"cny":8.6893,"hrk":9.07598,"ltl":19.7018,"mdl":3.28485,"kzt":0.16938199999999998,"aed":15.7586,"hnl":2.47562,"amd":0.120974,"etb":2.46175,"mvr":3.74848,"qar":15.8296,"sar":15.4332,"lbp":0.0383272,"bzd":28.9406,"bob":8.37907,"crc":0.101296,"xcd":21.4371,"nzd":41.4107,"bif":0.0332641,"dop":1.21545,"czk":2.61026}`;
const rates: ICurrencyRates = JSON.parse(ratesText);


describe('DynaCurrencies, converters', () => {
  it('should load the rates', () => {
    const dynaCurrencies: DynaCurrencies = new DynaCurrencies();
    expect(dynaCurrencies.hasRates).toBe(false);
    dynaCurrencies.updateRates(rates);
    expect(dynaCurrencies.count).not.toBe(0);
    expect(dynaCurrencies.hasRates).toBe(true);
  });

  it('should reset the rates', () => {
    const dynaCurrencies: DynaCurrencies = new DynaCurrencies();
    expect(dynaCurrencies.hasRates).toBe(false);
    expect(dynaCurrencies.lastUpdate).toBe(null);
    dynaCurrencies.updateRates(rates);
    expect(dynaCurrencies.count).not.toBe(0);
    expect(dynaCurrencies.hasRates).toBe(true);
    expect(dynaCurrencies.lastUpdate).not.toBe(null);
    dynaCurrencies.clearRates();
    expect(dynaCurrencies.count).toBe(0);
    expect(dynaCurrencies.hasRates).toBe(false);
    expect(dynaCurrencies.lastUpdate).toBe(null);
  });

  it('should generate a label with values', () => {
    const dynaCurrencies: DynaCurrencies = new DynaCurrencies();
    dynaCurrencies.updateRates(rates);
    const label = dynaCurrencies.convertToLabel(1, 'euR', 'Gbp');
    expect(label).not.toBe(null);
    if (!label) return;
    expect(label.text).not.toBe(null);
    expect(label.text).not.toBe(undefined);
    expect(label.text).not.toBe('');
    expect(label.values.currencyCode).not.toBe(undefined);
    expect(label.values.currencyName).not.toBe(undefined);
    expect(label.values.currencyNamePlural).not.toBe(undefined);
    expect(label.values.currencySymbol).not.toBe(undefined);
    expect(label.values.currencySymbolNative).not.toBe(undefined);
    expect(label.values.decimals >= 0).toBe(true);
    expect(label.values.value >= 0).toBe(true);
  });

  it('should get all currencies', () => {
    const dynaCurrencies: DynaCurrencies = new DynaCurrencies();
    dynaCurrencies.updateRates(rates);
    const allCurrencies: ICurrency[] = dynaCurrencies.getCurrencyRatesArray();
    expect(allCurrencies.length > 100).toBe(true);
    expect(allCurrencies.filter(c => c.symbol.indexOf('$') > -1).length > 20).toBe(true);
  });

  it('should get currencies by country code', () => {
    const dynaCurrencies: DynaCurrencies = new DynaCurrencies();
    dynaCurrencies.updateRates(rates);

    expect(dynaCurrencies.getCurrencyByCountry('us')).toMatchSnapshot('us');
    expect(dynaCurrencies.getCurrencyByCountry('fr')).toMatchSnapshot('fr');
    expect(dynaCurrencies.getCurrencyByCountry('gr')).toMatchSnapshot('gr');
    expect(dynaCurrencies.getCurrencyByCountry('tr')).toMatchSnapshot('tr');
    expect(dynaCurrencies.getCurrencyByCountry('hk')).toMatchSnapshot('hk');
    expect(dynaCurrencies.getCurrencyByCountry('unknown')).toMatchSnapshot('unknown');
  });

  it('should not convert when no rates', () => {
    const dynaCurrencies: DynaCurrencies = new DynaCurrencies();
    expect(dynaCurrencies.convert(3.12, 'eur', 'gbp')).toBe(null);
    dynaCurrencies.updateRates(rates);
    expect(dynaCurrencies.convert(3.12, 'eur', 'gbp')).not.toBe(null);
  });

  it('should convert DynaPrices', () => {
    const dynaCurrencies: DynaCurrencies = new DynaCurrencies();
    dynaCurrencies.updateRates(rates);
    const gbpDynaPrice = dynaCurrencies.convertDynaPrice(
      {
        value: 3.12,
        currency: 'euR',
      },
      'gbP',
    );
    expect(gbpDynaPrice).toMatchSnapshot();
  });

  it('should not convert DynaPrices when no rates', () => {
    const dynaCurrencies: DynaCurrencies = new DynaCurrencies();
    const gbpDynaPrice = dynaCurrencies.convertDynaPrice(
      {
        value: 3.12,
        currency: 'euR',
      },
      'gbP',
    );
    expect(gbpDynaPrice).toMatchSnapshot();
  });

  it('should convert to label', () => {
    const dynaCurrencies: DynaCurrencies = new DynaCurrencies();
    dynaCurrencies.updateRates(rates);
    const labelPrice = dynaCurrencies.convertToLabel(3.12, 'euR', 'gbP');
    expect(labelPrice).toMatchSnapshot();
  });

  it('should not convert label when no rates', () => {
    const dynaCurrencies: DynaCurrencies = new DynaCurrencies();
    const labelPrice = dynaCurrencies.convertToLabel(3.12, 'euR', 'gbP');
    expect(labelPrice).toMatchSnapshot();
  });
});
