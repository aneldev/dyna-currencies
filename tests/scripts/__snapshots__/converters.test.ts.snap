// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynaCurrencies, converters should convert DynaPrices 1`] = `
Object {
  "currency": "gbp",
  "value": 2.7521418263204316,
}
`;

exports[`DynaCurrencies, converters should convert to label 1`] = `
Object {
  "text": "2.75£",
  "values": Object {
    "currencyCode": "GBP",
    "currencyName": "British Pound Sterling",
    "currencyNamePlural": "British pounds sterling",
    "currencySymbol": "£",
    "currencySymbolNative": "£",
    "decimals": 2,
    "value": 2.7521418263204316,
  },
}
`;

exports[`DynaCurrencies, converters should get currencies by country code: fr 1`] = `
Object {
  "code": "EUR",
  "decimalDigits": 2,
  "name": "Euro",
  "namePlural": "euros",
  "rounding": 0,
  "symbol": "€",
  "symbolNative": "€",
}
`;

exports[`DynaCurrencies, converters should get currencies by country code: gr 1`] = `
Object {
  "code": "EUR",
  "decimalDigits": 2,
  "name": "Euro",
  "namePlural": "euros",
  "rounding": 0,
  "symbol": "€",
  "symbolNative": "€",
}
`;

exports[`DynaCurrencies, converters should get currencies by country code: hk 1`] = `
Object {
  "code": "HKD",
  "decimalDigits": 2,
  "name": "Hong Kong Dollar",
  "namePlural": "Hong Kong dollars",
  "rounding": 0,
  "symbol": "HK$",
  "symbolNative": "$",
}
`;

exports[`DynaCurrencies, converters should get currencies by country code: tr 1`] = `
Object {
  "code": "TRY",
  "decimalDigits": 2,
  "name": "Turkish Lira",
  "namePlural": "Turkish Lira",
  "rounding": 0,
  "symbol": "TL",
  "symbolNative": "TL",
}
`;

exports[`DynaCurrencies, converters should get currencies by country code: unknown 1`] = `null`;

exports[`DynaCurrencies, converters should get currencies by country code: us 1`] = `
Object {
  "code": "USD",
  "decimalDigits": 2,
  "name": "US Dollar",
  "namePlural": "US dollars",
  "rounding": 0,
  "symbol": "$",
  "symbolNative": "$",
}
`;

exports[`DynaCurrencies, converters should not convert DynaPrices when no rates 1`] = `null`;

exports[`DynaCurrencies, converters should not convert label when no rates 1`] = `null`;
